### ISSUE TITLE
Specificity API

### ISSUE THREAD

**Issue by LeaVerou** [w3c/css-houdini-drafts#96]
Do we have an API for selector specificity, or at least comparing two selectors by specificity (since returning a number for specificity is hard)?
Since selectors are convenient for JS too, it makes sense that there will be data structures where selectors play an integral role and often conflicts need to be resolved.
I needed this recently and it was so tricky to implement that I gave up and went with source order.
I guess it could be either part of the parser API, or of the new CSSOM.

**Comment by FremyCompany** [comment]
I am strongly in favor of this. :+1: :100: 

I know we could do better than a simple "compare selector strings" api, and that some feels we should have real selector objects, but I feel like this is what we need right now. Objects are interesting to optimize performance, but librairies can optimize performance in other ways by using caching if we go with strings, and build their own selector datastore.

@LeaVerou Shameless plug, though: https://github.com/FremyCompany/css-grid-polyfill/blob/master/src/core/css-cascade.js#L19 (this is based on @tabatkins css parser's tokenizer, and works reliably well for the interoperable subset of css).

**Comment by LeaVerou** [comment]
Regarding returning a number, the problem is that there's no absolute value we can return, since the base is variable (the spec says "a large enough base"). If we fix the base, we limit the number of ids/classes/attribute selectors/pseudos a selector can contain.

**Comment by jtojnar** [comment]
Don’t return numeric value, return a tuple-like object (or an array) – comparing the tuples is then simply a matter of [lexicographical ordering](https://en.wikipedia.org/wiki/Lexicographical_order).

It seems like JavaScript even uses lexicographical order for arrays: `[1, 2, 3, 4] < [2, 0, 0, 0]`

**Comment by tabatkins** [comment]
There've been suggestions to add a `Selector` class, with methods similar to querySelector/etc on it.  The benefit of this is that the selector can be "compiled", similar to compiling a regex, for potentially quicker execution if it's going to be called repeatedly.

Such an object would be the obvious place to hang a `readonly attribute sequence<long> specificity` attribute.  And yeah, it would have to return a 3-tuple of values.

> It seems like JavaScript even uses lexicographical order for array

No, it just seems that way for simple examples. What's actually happening is that the arrays are being stringified, and then compared ASCII-betically.  For example, `[2] < [10]` returns false, because `"2" < "10"` is false.

**Comment by felixfbecker** [comment]
I was surprised to see that the css typed om spec did not include a selector class. Currently, APIs like `CSSStyleSheet.insertRule()` and `CSSStyleRule.selectorText` all represent selectors as strings. But especially with shadow roots, I see many use cases for manually constructing and inspecting these. For example, I may want to copy all rules for a `.btn` class in a stylesheet to a shadow root. For that, I currently need to filter rules by a regex on the selector.

**Comment by tabatkins** [comment]
Yup, Typed OM is currently limiting itself to property values, not stylesheets as a whole. It'll come later, I promise!
